Проект. Роутинг в React приложении. Реализация механизмов авторизации…

a1exTee opened the pull request:
… в React приложении.

tihonovalexey29-star committed:
Проект. Роутинг в React приложении. Реализация механизмов авторизации в React приложении.

gennady-bars commented on src/utils/data.jsx:
```diff
@@ -7,4 +7,212 @@ const response = (res) => {
 export const requestIngredients = async () => {
   const res = await fetch(`${apiUrl}ingredients`);
   return response(res);
```
`response`

**Названия** методов и функций нужно начинать **с глагола** в начальной форме (такой пункт есть в чек-листе). Это обычная международная практика, чтобы функция говорила, что она делает. **Все функции в проекте нужно исправить, если не начинаются с глагола** 

 `checkResponse`

gennady-bars replied:
**Можно лучше**

Можно сделать **универсальную** функцию запроса с **проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  function request(url, options) {
    // принимает два аргумента: урл и объект опций, как и `fetch`
    return fetch(url, options).then(checkResponse)
  }

```

И теперь просто нужно заменить все `fetch` на `request` и убрать дублирование проверки на `ok`. Все остальное будет без изменений. Код станет чище

И даже можно поместить внутрь `baseUrl`, чтобы не дублировать его в каждом запросе. Тогда нужно будет передавать просто `endpoint` в вызов

gennady-bars commented on src/utils/data.jsx:
```diff
@@ -7,4 +7,212 @@ const response = (res) => {
+}
+
+
+const checkResponse = (res) => {
```
Вот эта функция только нужна для проверки на `ok`


Нужно удалить все дублирования этой функции под названием `response`. Их 2 шт в коде в проекте

gennady-bars commented on src/services/actions/auth/auth.jsx:
```diff
@@ -0,0 +1,184 @@
+    dispatch({ type: GET_USER_REQUEST });
+    getUserFetch()
+      .then(res => {
+        if (res?.success) {
```
`        if (res?.success) {
`

**Можно лучше**


Если интересно, то можно создать **универсальную** функцию запроса `request` , которая будет содержать в себе уже **базовый урл**, проверку на **ok** и проверку на `success`, чтобы не дублировать их в каждом запросе.

```js

// 1 раз объявляем базовый урл
export const BASE_URL = "https://norma.nomoreparties.space/api/";

// создаем функцию проверки ответа на `ok`
const checkResponse = (res) => {
  if (res.ok) {
    return res.json();
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ошибка ${res.status}`);
};

// создаем функцию проверки на `success`
const checkSuccess = (res) => {
  if (res && res.success) {
    return res;
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ответ не success: ${res}`);
};

// создаем универсальную фукнцию запроса с проверкой ответа и `success`
// В вызов приходит `endpoint`(часть урла, которая идет после базового) и опции
const request = (endpoint, options) => {
  // а также в ней базовый урл сразу прописывается, чтобы не дублировать в каждом запросе
  return fetch(`${BASE_URL}${endpoint}`, options)
    .then(checkResponse)
    .then(checkSuccess);
};

```


И теперь не нужно будет дублировать все это в каждом запросе. Нужно просто заменить в запросах все `fetch` на `request`, удалив лишние проверки и базовый урл. Все остальное останется без изменений

Код станет намного чище. Будет мелкий аналог `axios`

Пример запроса ингредиентов

```js
const getIngredients =  () => request("ingredients");

```

В `get`-запросах даже не нужно указывать 2й параметр. Он получится `undefined`, что без проблем для `fetch` (который внутри).

gennady-bars commented on src/components/app/App.jsx:
```diff
@@ -1,57 +1,61 @@
-    return <h2>Ошибка</h2>;
-  }
+  /*useEffect(() => {
+    dispatch(getMenu());
```
Вот это тут нужно как раз для получения ингредиентов  1 раз. Нигде больше не нужно дублировать

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

К сожалению, работу придется отклонить без потери итерации, так как не выполнена часть задания:

* По заданию нужно сделать логику, при которой бы показывался попап с деталями ингредиента, если кликнуть по ингредиенту, но если скопировать ссылку на ингредиент при открытом попапе и **вставить в окно браузера**, то должна уже показываться **страница** с этими деталями. 
* Если **перезагрузить** страницу с открытым попапом ингредиента, то этот открытый попап опять должен появляться (или страница с деталями). А сейчас ничего не отображается после перезагрузки. 
* Вам нужно использовать средства **роутинга** для определения, показывать ли попап с деталями, или показывать страницу. Вам должны были дать ссылку на документацию по этой логике. Вот она https://v5.reactrouter.com/web/example/modal-gallery . Пожалуйста, посоветуйтесь с наставником. В 6й версии роутинга примерно такая же логика.




Еще заметил ошибки:


* Нужно удалить все дублирования этой функции под названием `response`. Их 2 шт в коде в проекте. Только `checkResponse ` нужно использовать в коде

tihonovalexey29-star committed:
убрал дубли функции response, при клике на ингредиет добавляется url с id

tihonovalexey29-star committed:
поправил замечание - Вот это тут нужно как раз для получения ингредиентов 1 раз. Нигде больше не нужно дублировать

tihonovalexey29-star committed:
добавил проверки на детальной ингредиента

tihonovalexey29-star committed:
Исправлено:
    По заданию нужно сделать логику, при которой бы показывался попап с деталями ингредиента, если кликнуть по ингредиенту, но если скопировать ссылку на ингредиент при открытом попапе и вставить в окно браузера, то должна уже показываться страница с этими деталями.
    Если перезагрузить страницу с открытым попапом ингредиента, то этот открытый попап опять должен появляться (или страница с деталями). А сейчас ничего не отображается после перезагрузки.

tihonovalexey29-star committed:
fix кнопки Оформить заказ

gennady-bars commented on src/components/form/Form.jsx:
```diff
@@ -0,0 +1,39 @@
+    setValues({...values, [name]: value});
+  };
+  return {values, setValues, handleChange };
+};
```
Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

gennady-bars commented on src/components/form/Form.jsx:
```diff
@@ -0,0 +1,39 @@
+      ? <h3 className={`text text_type_main-medium ${formStyles.title}`}>{title}</h3>
+      : null
+      }
+      <form onSubmit={onSubmit} className={formStyles.form}>
```
Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. Обработка события `submit` еще срабатывает при нажатии на `Enter`, так как это встроенная логика сабмитов в браузере.

gennady-bars commented on src/pages/HomePage/homePage.jsx:
```diff
@@ -0,0 +1,48 @@
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    dispatch(getIngredients());
```
`dispatch(getIngredients())` достаточно 1 раз сделать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/pages/IngredientDetailsPage/IngredientDetailsPage.jsx:
```diff
@@ -0,0 +1,42 @@
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    dispatch(getIngredients());
```
`dispatch(getIngredients())` достаточно 1 раз сделать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/components/ProtectedRoute.jsx:
```diff
@@ -0,0 +1,19 @@
+  const isAuthorized = useSelector((store) => store.authReducer.isAuthorized);
+
+  const location = useLocation();
+  const from = location.state?.from || '/';
```
Хорошо сделан `ProtectedRoute`

gennady-bars commented on src/components/app/App.jsx:
```diff
@@ -1,57 +1,73 @@
   const dispatch = useDispatch();
+  const location = useLocation();
+  const navigate = useNavigate();
+  const background = location.state && location.state.background;
```
Отлично сделана логика проверки, был ли пользователь на сайте, и показывать ли страницу  с деталями, а не модально окно

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку
* Хорошо сделан `ProtectedRoute`
* Отлично сделана логика проверки, был ли пользователь на сайте, и показывать ли страницу  с деталями, а не модально окно


но есть некоторые недочеты:




* Приложение ломается при перезагрузке страницы с открытым попапом https://disk.yandex.ru/i/U-celuGErh7OBA 
* Код хука нужно поместить в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты
* `dispatch(getIngredients())` достаточно 1 раз сделать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.



**Можно лучше**

* лучше отображать **лоадер** (загрузку) при ожидании ответа от сервера с номером заказа



Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

tihonovalexey29-star committed:
Исправлено:
    Приложение ломается при перезагрузке страницы с открытым попапом https://disk.yandex.ru/i/U-celuGErh7OBA
    Код хука нужно поместить в отдельный файл useForm.js в папке hooks и импортируют функцию туда, где нужно контролировать инпуты
    dispatch(getIngredients()) достаточно 1 раз сделать в App, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
